### PR TITLE
Stackalloc and marshal the delivery tag

### PR DESCRIPTION
**Pull request by danielmarbach**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs [hunk 1]
```diff
@@ -360,9 +362,13 @@ public static ServiceBusReceivedMessage AmqpMessageToSBMessage(AmqpMessage amqpM
 
             if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
             {
-                var guidBuffer = new byte[GuidSizeInBytes];
-                Buffer.BlockCopy(amqpMessage.DeliveryTag.Array, amqpMessage.DeliveryTag.Offset, guidBuffer, 0, GuidSizeInBytes);
-                sbMessage.LockTokenGuid = new Guid(guidBuffer);
+                Span<byte> guidBytes = stackalloc byte[GuidSizeInBytes];
```
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:365]:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.201
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  Job-CHTIXI : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT

InvocationCount=9600000  

```
|               Method |      Mean |     Error |    StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------:|----------:|----------:|------:|-------:|------:|------:|----------:|
|   BufferAndBlockCopy | 11.426 ns | 0.0638 ns | 0.0565 ns |  1.00 | 0.0047 |     - |     - |      40 B |
| SpanAndMemoryMarshal |  9.899 ns | 0.0271 ns | 0.0212 ns |  0.87 |      - |     - |     - |         - |

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:365]:
```
namespace MicroBenchmarks.ServiceBus
{
    using System;
    using System.Runtime.InteropServices;
    using BenchmarkDotNet.Attributes;
    using BenchmarkDotNet.Configs;
    using BenchmarkDotNet.Diagnosers;
    using BenchmarkDotNet.Exporters;
    using BenchmarkDotNet.Jobs;

    [Config(typeof(Config))]
    public class GuidConversation
    {
        private ArraySegment<byte> data;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(9600000));
            }
        }
        [IterationSetup]
        public void Setup()
        {
            data = new ArraySegment<byte>(Guid.NewGuid().ToByteArray());
        }

        [Benchmark(Baseline = true)]
        public Guid BufferAndBlockCopy()
        {
            var guidBuffer = new byte[16];
            Buffer.BlockCopy(data.Array, data.Offset, guidBuffer, 0, 16);
            return new Guid(guidBuffer);
        }

        [Benchmark]
        public Guid SpanAndMemoryMarshal()
        {
            Span<byte> guidBytes = stackalloc byte[16];
            data.AsSpan().CopyTo(guidBytes);
            if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
            {
                lockTokenGuid = new Guid(guidBytes.ToArray());
            }
            return lockTokenGuid;
        }
    }
}
```

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs [hunk 2]
```diff
@@ -360,9 +362,13 @@ public static ServiceBusReceivedMessage AmqpMessageToSBMessage(AmqpMessage amqpM
...
+                amqpMessage.DeliveryTag.AsSpan().CopyTo(guidBytes);
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:366]: I may be overlooking something important about the raw form of the AMQP buffers, but I'm wondering why we need to make the intermediate copy.  

There is a `Guid` constructor that takes a `ReadOnlySpan<byte>` so we know it's not attempting to mutate the argument.  Would following would work?

```csharp
if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
{
    sbMessage.LockTokenGuid= new Guid(amqpMessage.DeliveryTag.AsSpan());
}
```

If I'm not accounting for some oddity that `MemoryMarshal` handles, maybe a fallback case?

```csharp
if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
{
    if (!MemoryMarshal.TryRead<Guid>(amqpMessage.DeliveryTag.AsSpan(), out var lockTokenGuid))
    {
        lockTokenGuid = new Guid(amqpMessage.DeliveryTag.Array);
    }

    sbMessage.LockTokenGuid = lockTokenGuid;
}
```

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:366]: The problem is netstandard. We can't properly multi target here to pick the appropriate guid overload. Or at least I tried and failed

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:366]: If we'd dual target netstandard 2 and 2.1 your proposal would work but that would be a whole other story I guess

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:366]: ah, thanks.  That's my bad.  I figured that I was overlooking something.  My test app was defaulted to .NET 5.

**Review by jsquire — APPROVED** [review]

**Pull request merged**